internal/resources/google/artifact_registry_repository.go

```diff
@@ -0,0 +1,368 @@
+package google
+
+import (
+	"fmt"
+	"github.com/infracost/infracost/internal/resources"
+	"github.com/infracost/infracost/internal/schema"
+	"github.com/infracost/infracost/internal/usage"
+	"github.com/shopspring/decimal"
+	"reflect"
+	"regexp"
+	"strings"
+)
+
+var (
+	artifactRegSvcName = strPtr("Artifact Registry")
+
+	continentDefault      = "Intercontinental (Excl Oceania)"
+	continentNorthAmerica = "North America"
+	continentSouthAmerica = "South America"
+	continentEurope       = "Europe"
+	// GCP misspells AsiaPacific without the i. The below is not a typo!
+	continentApac    = "AsiaPacfic"
+	continentOceania = "Oceania"
+
+	artifactGlobalEgressContinents = map[string]struct{}{
+		continentOceania:      {},
+		continentSouthAmerica: {},
+	}
+	artifactRegMultiRegionNames = map[string]struct{}{
+		"us":     {},
+		"europe": {},
+		"asia":   {},
+	}
+
+	regionSep = regexp.MustCompile(`[_\-]`)
+)
+
+// ArtifactRegistryRepository struct represents a GCP artifact Registry. Artifact registry is essentially
+// a next generation version of google's container registry. It allows users to store container images and language
+// packages in the GCP.
+//
+// Pricing for Artifact Registry is based on storage amounts and data transfer.
+//
+// Resource information: https://cloud.google.com/artifact-registry
+// Pricing information: https://cloud.google.com/artifact-registry/pricing
+type ArtifactRegistryRepository struct {
+	Address   string
+	Region    string
+	Continent string
+
+	// StorageGB represents a usage cost that defines the amount of gb the artifact registry uses on a per monthly basis.
+	StorageGB *float64 `infracost_usage:"storage_gb"`
+	// MonthlyEgressDataTransferGB represents a complex usage cost that defines data transfer to different regions in the
+	// google cloud infra. This does not include outbound internet egress (e.g. downloading artifact data to a local machine).
+	MonthlyEgressDataTransferGB *locationDataTransfer `infracost_usage:"monthly_egress_data_transfer_gb"`
+}
+
+// locationDataTransfer represents a usage map that allows the users to specify which regions the artifact registry
+// has data egress to/from.
+type locationDataTransfer struct {
```

Comment:
yep I'll change up north america and us naming. This just got lost when I was renaming the usage costs from their struct tags using multiline carets....

I am using a struct approach here but because I want to loop over it I convert it to a slice with specific keys. It'll be easier to discuss on a call tomorrow as I'm not sold on this approach as well. Chat in the morning.

Reply:
changed naming here: https://github.com/infracost/infracost/pull/1203/commits/7f8b99304acefa63780d0b196b74a6e5c2d714cc & https://github.com/infracost/infracost/pull/1203/commits/e47ac6537708db923bc02996182169ddb08e213f